syft/pkg/cataloger/terraform/parse_tf_lock.go

```diff
@@ -0,0 +1,55 @@
+package terraform
+
+import (
+	"context"
+	"fmt"
+	"io"
+
+	"github.com/anchore/packageurl-go"
+	"github.com/hashicorp/hcl/v2/hclsimple"
+
+	"github.com/anchore/syft/syft/artifact"
+	"github.com/anchore/syft/syft/file"
+	"github.com/anchore/syft/syft/pkg"
+	"github.com/anchore/syft/syft/pkg/cataloger/generic"
+)
+
+type terraformLockFile struct {
+	Providers []pkg.TerraformLockEntry `hcl:"provider,block"`
+}
+
+func parseTerraformLock(_ context.Context, _ file.Resolver, _ *generic.Environment, reader file.LocationReadCloser) ([]pkg.Package, []artifact.Relationship, error) {
+	var lockFile terraformLockFile
+
+	contents, err := io.ReadAll(reader)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to read terraform lock file: %w", err)
+	}
+
+	err = hclsimple.Decode(reader.RealPath, contents, nil, &lockFile)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to decode terraform lock file: %w", err)
+	}
+
+	pkgs := make([]pkg.Package, 0, len(lockFile.Providers))
+
+	for _, provider := range lockFile.Providers {
+		pkg := pkg.Package{
+			Name:      provider.URL,
+			Version:   provider.Version,
+			FoundBy:   "terraform-cataloger",
+			Locations: file.NewLocationSet(reader.Location.WithAnnotation(pkg.EvidenceAnnotationKey, pkg.PrimaryEvidenceAnnotation)),
+			Licenses:  pkg.NewLicenseSet(), // TODO: license could be found in .terraform/providers/${name}/${version}/${arch}/LICENSE.txt
+			// TODO: Language?
+			Type: pkg.TerraformPkg,
+			// TODO: CPEs?
```

Comment:
I don't think we should generate CPEs for terraform providers quite yet, since it does not have a clear way to be paired up with NVD vulnerability data and that is the primary purpose for this. If there was an obviously correct way to make a bullet proof CPE from the terraform data then I'd say go for it now in this PR, but I don't think that's quite the case so I think the aforementioned point hints to leave it out for now.

Reply:
I agree 👍🏻